lib/request_overrider.js

```diff
@@ -625,28 +616,35 @@ function parseFullReplyResult(response, fullReplyResult) {
   }
 
   response.statusCode = status
+  response.rawHeaders.push(...common.headersInputToRawArray(headers))
+  debug('response.rawHeaders after reply: %j', response.rawHeaders)
 
-  if (headers) {
-    debug('new headers: %j', headers)
+  return body
+}
 
-    const rawHeaders = Array.isArray(headers)
-      ? headers
-      : [].concat(...Object.entries(headers))
-    response.rawHeaders.push(...rawHeaders)
+/**
+ * Determine which of the default headers should be added to the response.
+ *
+ * Don't include any defaults whose case-insensitive keys are already on the response.
+ */
+function selectDefaultHeaders(existingHeaders, defaultHeaders) {
+  if (!defaultHeaders.length) {
+    return [] // return early if we don't need to bother
+  }
 
-    const castHeaders = Array.isArray(headers)
-      ? common.headersArrayToObject(headers)
-      : common.headersFieldNamesToLowerCase(headers)
+  const definedHeaders = new Set()
+  const result = []
 
-    response.headers = {
-      ...response.headers,
-      ...castHeaders,
+  for (let i = 0, len = existingHeaders.length; i < len; i = i + 2) {
```

Comment:
Maybe it would be helpful to define `forEachHeaderValue` that invokes a callback with `(value, key, indexOfKey)` parameters? It could be used in the three places here where these headers are being iterated.